ignore ill-formed solution of reduce_inequalities

avikchaudhuri opened the pull request:
Summary:
Fixes https://github.com/pytorch/pytorch/issues/117033

Sometimes the solution returned by `sympy.solvers.inequalities.reduce_inequalities` can contain sub-expressions of the form `CRootOf(...)`, denoting the complex root of some equation in `x`, where `x` is an arbitrary symbol. We will now gracefully fail when this happens, like we already do when the solver itself fails.

Test Plan: added a test

Differential Revision: D52715578

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1405,7 +1407,7 @@ def solve(self, disable_congruences=True, disable_equivalences=True):
                 else:
                     self._dynamic_results.add(self._dcp.doprint(solution))
-            except NotImplementedError as e:
+            except (NotImplementedError, AssertionError) as e:
```
meh, not a huge fan of catching all assertion errors (you're liable to catch legit assertion problems) but I guess it's fine

Flamefire replied:
I agree too that "AssertionError" is semantically wrong here.   
Especially as next to the intended assertion there is another assertion that really is an assertion, i.e. something that should not fail.

avikchaudhuri committed:
ignore ill-formed solution of reduce_inequalities (#117310)

Summary:

Fixes https://github.com/pytorch/pytorch/issues/117033

Sometimes the solution returned by `sympy.solvers.inequalities.reduce_inequalities` can contain sub-expressions of the form `CRootOf(...)`, denoting the complex root of some equation in `x`, where `x` is an arbitrary symbol. We will now gracefully fail when this happens, like we already do when the solver itself fails.

Test Plan: added a test

Differential Revision: D52715578

avikchaudhuri commented:
@pytorchbot merge

Flamefire commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1116,7 +1116,9 @@ def print_source(self, source) -> str:
     def _print_Symbol(self, expr) -> str:
         assert isinstance(expr, sympy.Symbol), str(type(expr))
-
+        assert self.symbol_to_source.get(expr), (
+            f"Unknown symbol {expr} created by constraints solver"
+        )
         return self.print_source(self.symbol_to_source[expr][0])
```
This is not optimal as you are doing a dict-lookup twice. Also python allows to remove all assertions: `python -O test.py`

What about:
source = self.symbol_to_source.get(expr)
```
if source:
  return source[0]
else:
  raise UnknownSymbolError(f"Unknown symbol {expr} created by constraints solver")
```

avikchaudhuri replied:
The surrounding code has similar assertions (see `ShapeGuardPrinter`). Will stamp a PR if you want to put one up to upgrade all of them.

Flamefire replied:
You are referring to https://github.com/avikchaudhuri/pytorch/blob/a10a820e707191cf9bc75275841a05eabdec106e/torch/fx/experimental/symbolic_shapes.py#L1086-L1090 ?

The difference is: That `assert` looks like a "real assertion", i.e. something that should hold true and is a bug if it does not. (But TBH I don't understand the text, i.e. what the issue in that case is)   
While here you create an exception which is intended to be handled at the call site. And as @ezyang mentioned with this you can't differentiate between those and might wrongly ignore a valid one.

asmeurer commented:
Just to be clear, the CRootOf objects should represent real numbers. SymPy should probably be updated here to simplify the result of reduce_inequalities when the variable is an integer, since something like `s0 > expr` can be simplified to `s0 > int(expr)`.

asmeurer commented:
I opened https://github.com/sympy/sympy/issues/26069 to track it.
